tests/storage/coalesce/test_coalesce.py

```diff
@@ -0,0 +1,42 @@
+import pytest
+
+import logging
+
+from .utils import compare_data, copy_data_to_tapdev, operation_on_vdi, wait_for_vdi_coalesce
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from lib.host import Host
+    from lib.vdi import VDI
+
+class Test:
+    def test_write_data(self, host: "Host", tapdev: str, data_file_on_host: str):
+        length = 1 * 1024 * 1024
+        offset = 0
+
+        logging.info("Copying data to tapdev")
+        copy_data_to_tapdev(host, data_file_on_host, tapdev, offset, length)
+
+        logging.info("Comparing data to tapdev")
+        assert compare_data(host, tapdev, data_file_on_host, offset, length)
+
+    @pytest.mark.parametrize("vdi_op", ["snapshot", "clone"])
+    def test_coalesce(self, host: "Host", tapdev: str, vdi_with_vbd_on_dom0: "VDI", data_file_on_host: str, vdi_op):
```

Comment:
Having to use `vdi_with_vbd_on_dom0` to get access to the VDI behind `tapdev` is confusing.
Maybe a `TapDev` object to encapsulate all relevant info would improve this?